nameko/events.py

```diff
@@ -183,10 +168,63 @@ def __init__(self, source_service, event_type, handler_type=SERVICE_POOL,
         super(EventHandler, self).__init__(
             queue=None, requeue_on_error=requeue_on_error)
 
+    @property
+    def broadcast_identifier(self):
+        """ A unique string to identify a service instance for `BROADCAST`
+        type handlers.
+
+        The `broadcast_identifier` is appended to the queue name when the
+        `BROADCAST` handler type is used. It must uniquely identify service
+        instances that receive broadcasts.
+
+        The default `broadcast_identifier` is a uuid that is set when the
+        service starts. It will change when the service restarts, meaning
+        that any unconsumed messages that were broadcast to the 'old' service
+        instance will not be received by the 'new' one. ::
+
+            @property
+            def broadcast_identifier(self):
+                # use a uuid as the identifier.
+                # the identifier will change when the service restarts and
+                # any unconsumed messages will be lost
+                return uuid.uuid4().hex
+
+        The default behaviour is therefore incompatible with reliable delivery.
+
+        An alternative `broadcast_identifier` that would survive service
+        restarts is ::
+
+            @property
+            def broadcast_identifier(self):
+                # use the machine hostname as the identifier.
+                # this assumes that only one instance of a service runs on
+                # any given machine
+                return socket.gethostname()
+
+        If neither of these approaches are appropriate, you could read the
+        value out of a configuration file ::
+
+            @property
+            def broadcast_identifier(self):
+                return self.config['SERVICE_IDENTIFIER']  # or similar
+
+        Broadcast queues are exclusive to ensure that `broadcast_identifier`
+        values are unique.
+        """
+        if self.reliable_delivery and self.handler_type is BROADCAST:
```

Comment:
I see your point about sharing a class for "service pool" and "broadcast" events. It's that way because they're (even now) very similar in implementation. It is only the queue name, and now the `exclusive` value, that changes.

I would advocate a separate class if it helped from an API perspective, but I don't think it does. Broadcast makes sense to me as an extension of the default "service pool" behaviour; they're both "event handlers"; if we made them separate implementations we'd need to give them distinct names, and I can't think of good ones.

If a single implementation with multiple modes is appropriate then I don't think it's "bad" in any sense to have an attribute, property or method that doesn't always apply. As you say, it's fine to have undefined behaviour when calling "out of turn".

I have converted `broadcast_handler` to be method rather than a property, for consideration, but I don't think it's really an improvement. It feels less idiomatic without benefit to me.

With regard to `getmembers` and descriptors... Is it really better to say "don't use descriptors" than it is to say "make sure descriptors are safe to be invoked at any time"? I think latter is more pythonic.